### PR TITLE
Use split queries on object caching is enabled.

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/57296

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-query.php [hunk 1]
```diff
@@ -3221,7 +3221,7 @@ public function get_posts() {
 		}
 
 		if ( null === $this->posts ) {
-			$split_the_query = ( $old_request == $this->request && "{$wpdb->posts}.*" === $fields && ! empty( $limits ) && $q['posts_per_page'] < 500 );
+			$split_the_query = ( $old_request == $this->request && "{$wpdb->posts}.*" === $fields && ( wp_using_ext_object_cache() || ( ! empty( $limits ) && $q['posts_per_page'] < 500 ) ) );
```
Comment (SergeyBiryukov) [inline src/wp-includes/class-wp-query.php]:
```suggestion
			$split_the_query = (
				$old_request == $this->request
				&& "{$wpdb->posts}.*" === $fields
				&& (
					wp_using_ext_object_cache()
					|| ( ! empty( $limits ) && $q['posts_per_page'] < 500 )
				)
			);
```
This conditional is super hard to read, let's split it into shorter lines 🙂

**Review by peterwilsoncc — COMMENTED** [review]
Some tests would be lovely to ensure that the change has the desired effect. There will need to be an assertion that expects different results on the runs with memcached. 

There should probably be some tests with the other conditions too, I can't see any at a quick glance.

**Comment by peterwilsoncc** [comment]
The unit test failures in `test_wp_get_nav_menu_items_cache_primes_posts` on memcached configs look legitimate. I guess because the query used in those tests are now split.

**Comment by spacedmonkey** [comment]
@peterwilsoncc Unit tests are fixed. The unit test needed to updated, as there is a change in behaviour.

File: tests/phpunit/tests/post/query.php [hunk 2]
```diff
@@ -775,4 +774,21 @@ public function test_found_posts_should_be_integer_even_if_found_posts_filter_re
...
+		$filter = new MockAction();
+		add_filter( 'split_the_query', array( $filter, 'filter' ) );
+
+		$q = new WP_Query(
+			array(
+				'posts_per_page' => 501,
+			)
+		);
+
+		$this->assertSame( (bool) wp_using_ext_object_cache(), $filter->get_args()[0][0] );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/query.php:792]: This is a really nice test. :)

**Review by peterwilsoncc — APPROVED** [review]
LGTM, thanks @spacedmonkey

**Comment by spacedmonkey** [comment]
Committed https://core.trac.wordpress.org/changeset/56513

**Pull request closed without merging**